spec/lib/saml_idp/request_spec.rb

```diff
@@ -479,6 +479,146 @@ module SamlIdp
       end
     end
 
+    describe '#cert_errors' do
+      let(:saml_request) { make_saml_request }
+
+      subject do
+        described_class.from_deflated_request saml_request
+      end
+
+      describe 'document is not signed' do
+        it 'returns nil' do
+          expect(subject.cert_errors).to be_nil
+        end
+      end
+
+      describe 'document is signed' do
+        let(:saml_request) { signed_auth_request }
+        let(:service_provider) { subject.service_provider }
+        let(:cert) { saml_settings.get_sp_cert }
+
+        describe 'the service provider has no registered certs' do
+          before { subject.service_provider.certs = [] }
+
+
+          it 'returns a no registered cert error' do
+            expect(subject.cert_errors).to eq [{cert: nil, error_code: :no_registered_certs}]
+          end
+        end
+
+        describe 'the service provider has one registered cert' do
+          before { subject.service_provider.certs = [cert] }
+          let(:errors) { [{ cert: cert.serial.to_s, error_code: error_code }] }
+
+          describe 'the cert matches the assertion cert' do
+            it 'returns nil' do
+              expect(subject.cert_errors).to be_nil
+            end
+          end
+
+          describe 'the embedded certificate is bad' do
+            let(:error_code) { :invalid_certificate_in_request }
+            let(:cert_text) do
+              invalid_cert.to_s.
+              gsub('-----BEGIN CERTIFICATE-----', '').
+              gsub('-----END CERTIFICATE-----', '').
+              gsub("\n", '')
+            end
+            before do
+              allow(OpenSSL::X509::Certificate).to receive(:new).with(Base64.decode64(cert_text)).and_raise OpenSSL::X509::CertificateError
```

Comment:
is it possible to just pass in a bad cert?

```ruby
> OpenSSL::X509::Certificate.new('aaaa')
(irb):2:in `initialize': PEM_read_bio_X509: no start line (OpenSSL::X509::CertificateError)
```

Reply:
unfortunately not. i originally tried that, but we depend on `ruby-saml` for the request set up, and if you try to pass a bad certificate into that, it fails for exactly that reason

Reply:
bummer, thanks for confirming!